components/eam/src/physics/cam/hetfrz_classnuc_cam.F90

```diff
@@ -539,114 +513,129 @@ end subroutine hetfrz_classnuc_cam_init
    enddo
 
-   ! Convert interstitial and cloud borne aerosols from a mass to a volume basis before
+   ! get interstitial aerosols from state_q
+   ! Convert interstitial aerosols from a mass to a volume basis before
    ! being used in get_aer_num
-   do ispec = 1, ncnst
-      aer_cb(:ncol,:,ispec,lchnk_zb) = aer_cb(:ncol,:,ispec,lchnk_zb) * rho(:ncol,:)
-
-      ! Check whether constituent is a mass or number mixing ratio
-      if (spec_idx(ispec) == 0) then
-         call rad_cnst_get_mode_num(0, mode_idx(ispec), 'a', state, pbuf, ptr2d)
-      else
-         call rad_cnst_get_aer_mmr(0, mode_idx(ispec), spec_idx(ispec), 'a', state, pbuf, ptr2d)
-      endif
-      aer(:ncol,:,ispec,lchnk_zb) = ptr2d(:ncol,:) * rho(:ncol,:)
-   enddo
+
+   aer(:ncol,:pver,so4_accum,lchnk_zb)  = state_q(:ncol,:pver,lptr_so4_a_amode(modeptr_accum)) * rho(:ncol,:)
+   aer(:ncol,:pver,bc_accum, lchnk_zb)  = state_q(:ncol,:pver,lptr_bc_a_amode(modeptr_accum)) * rho(:ncol,:)
+   aer(:ncol,:pver,pom_accum,lchnk_zb)  = state_q(:ncol,:pver,lptr_pom_a_amode(modeptr_accum)) * rho(:ncol,:)   
+   aer(:ncol,:pver,soa_accum,lchnk_zb)  = state_q(:ncol,:pver,lptr_soa_a_amode(modeptr_accum)) * rho(:ncol,:)
+   aer(:ncol,:pver,dst_accum,lchnk_zb)  = state_q(:ncol,:pver,lptr_dust_a_amode(modeptr_accum)) * rho(:ncol,:)
+   aer(:ncol,:pver,ncl_accum,lchnk_zb)  = state_q(:ncol,:pver,lptr_nacl_a_amode(modeptr_accum)) * rho(:ncol,:)
+   aer(:ncol,:pver,mom_accum,lchnk_zb)  = state_q(:ncol,:pver,lptr_mom_a_amode(modeptr_accum)) * rho(:ncol,:)
+   aer(:ncol,:pver,num_accum,lchnk_zb)  = state_q(:ncol,:pver,numptr_amode(modeptr_accum)) * rho(:ncol,:)
+
+   aer(:ncol,:pver,dst_coarse,lchnk_zb) = state_q(:ncol,:pver,lptr_dust_a_amode(modeptr_coarse)) * rho(:ncol,:)
+   aer(:ncol,:pver,ncl_coarse,lchnk_zb) = state_q(:ncol,:pver,lptr_nacl_a_amode(modeptr_coarse)) * rho(:ncol,:)
+   aer(:ncol,:pver,so4_coarse,lchnk_zb) = state_q(:ncol,:pver,lptr_so4_a_amode(modeptr_coarse)) * rho(:ncol,:)
+   aer(:ncol,:pver,bc_coarse, lchnk_zb) = state_q(:ncol,:pver,lptr_bc_a_amode(modeptr_coarse)) * rho(:ncol,:)
+   aer(:ncol,:pver,pom_coarse,lchnk_zb) = state_q(:ncol,:pver,lptr_pom_a_amode(modeptr_coarse)) * rho(:ncol,:)
+   aer(:ncol,:pver,soa_coarse,lchnk_zb) = state_q(:ncol,:pver,lptr_soa_a_amode(modeptr_coarse)) * rho(:ncol,:)
+   aer(:ncol,:pver,mom_coarse,lchnk_zb) = state_q(:ncol,:pver,lptr_mom_a_amode(modeptr_coarse)) * rho(:ncol,:)
+   aer(:ncol,:pver,num_coarse,lchnk_zb) = state_q(:ncol,:pver,numptr_amode(modeptr_coarse)) * rho(:ncol,:)
+
+   aer(:ncol,:pver,bc_pcarbon, lchnk_zb) = state_q(:ncol,:pver,lptr_bc_a_amode(modeptr_pcarbon)) * rho(:ncol,:)
+   aer(:ncol,:pver,pom_pcarbon,lchnk_zb) = state_q(:ncol,:pver,lptr_pom_a_amode(modeptr_pcarbon)) * rho(:ncol,:)
+   aer(:ncol,:pver,mom_pcarbon,lchnk_zb) = state_q(:ncol,:pver,lptr_mom_a_amode(modeptr_pcarbon)) * rho(:ncol,:)
+   aer(:ncol,:pver,num_pcarbon,lchnk_zb) = state_q(:ncol,:pver,numptr_amode(modeptr_pcarbon)) * rho(:ncol,:)
```

Comment:
This single array `aer` with a combination of aerosol mass and number seems like a bit of a mess (@mingxuanwupnnl not blaming you for this). Can't we refactor this to use the more typical indexing fundamental to MAM? @jeff-cohere and I talked about this array and the associated indexing and agreed it won't be replicated in the C++ code.

Reply:
This process pick and chooses only some species from each mode. I didn't find any pattern to extract these automatically (maybe I should look harder). It is more or less hardwired for each MAM version (MAM3, 4, and 7). Maybe we can specify this info/order only once and use that in the rest of the code so that we have only one place to change if something changes in the future.